grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/autoconfigure/metrics/GRpcMetricsAutoConfiguration.java

```diff
@@ -88,39 +82,75 @@ protected MonitoringServerCall(ServerCall<ReqT, RespT> delegate, MeterRegistry r
+                closed = true;
                 final Timer.Builder timerBuilder = Timer.builder("grpc.server.calls");
-                tagsContributors.forEach(c->
-                        timerBuilder.tags(c.getTags(status,getMethodDescriptor(),getAttributes()))
+                tagsContributors.forEach(c ->
+                    timerBuilder.tags(c.getTags(status, getMethodDescriptor(), getAttributes()))
                 );
                 Optional.ofNullable(additionalTags)
-                        .ifPresent(timerBuilder::tags);
-
+                    .ifPresent(timerBuilder::tags);
                 start.stop(timerBuilder.register(registry));
             }
-
-
             super.close(status, trailers);
         }
 
-        public void addTags(List<Tag> tags) {
-            additionalTags  = tags;
+        public void addTags(Iterable<Tag> tags) {
+            additionalTags = additionalTags.and(tags);
         }
     }
 
     @Slf4j
-    static class MonitoringServerInterceptor implements ServerInterceptor, Ordered {
-
-        private MeterRegistry registry;
+    private static class MessageMonitoringListener<ReqT> extends ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT> {
+        private final MonitoringServerCall<ReqT, ?> monitoringServerCall;
+        private final Collection<RequestAwareGRpcMetricsTagsContributor<?>> requestAwareContributorCandidates;
+
+        protected MessageMonitoringListener(
+            ServerCall.Listener<ReqT> delegate,
+            MonitoringServerCall<ReqT, ?> monitoringServerCall,
+            Collection<RequestAwareGRpcMetricsTagsContributor<?>> requestAwareContributorCandidates
+        ) {
+            super(delegate);
+            this.monitoringServerCall = monitoringServerCall;
+            this.requestAwareContributorCandidates = requestAwareContributorCandidates;
+        }
 
+        @Override
+        public void onMessage(ReqT message) {
+            for (RequestAwareGRpcMetricsTagsContributor<?> contributor : requestAwareContributorCandidates) {
+                if (contributor.accepts(message)) {
+                    try {
+                        //noinspection unchecked
+                        monitoringServerCall.addTags(
+                            ((RequestAwareGRpcMetricsTagsContributor<ReqT>) contributor).addTags(
+                                message,
+                                monitoringServerCall.getMethodDescriptor(),
+                                monitoringServerCall.getAttributes(),
+                                monitoringServerCall.getAdditionalTags()
+                            )
+                        );
+                    } catch (RuntimeException error) {
+                        log.error("Failed to  execute tag contributor {}", contributor, error);
+                    }
+                }
+            }
+            super.onMessage(message);
+        }
```

Comment:
This is all safe, because:

> Implementations are not required to be thread-safe, but they must not be thread-hostile. The caller is free to call an instance from multiple threads, but only one call simultaneously.